firebase-firestore/src/main/java/com/google/firebase/firestore/pipeline/stage.kt

```diff
@@ -0,0 +1,671 @@
+ * of parameters accepted by the stage.
+ *
+ * This class provides a way to call stages that are supported by the Firestore backend but that are
+ * not implemented in the SDK version being used.
+ */
+class RawStage
+private constructor(
+  name: String,
+  private val arguments: List<GenericArg>,
+  options: InternalOptions = InternalOptions.EMPTY
+) : Stage<RawStage>(name, options) {
+  companion object {
+    /**
+     * Specify name of stage
+     *
+     * @param name The unique name of the stage to add.
+     * @return [RawStage] with specified parameters.
+     */
+    @JvmStatic fun ofName(name: String) = RawStage(name, emptyList(), InternalOptions.EMPTY)
+  }
+
+  override fun self(options: InternalOptions) = RawStage(name, arguments, options)
+
+  /**
+   * Specify arguments to stage.
+   *
+   * @param arguments A list of ordered parameters to configure the stage's behavior.
+   * @return [RawStage] with specified parameters.
+   */
+  fun withArguments(vararg arguments: Any): RawStage =
+    RawStage(name, arguments.map(GenericArg::from), options)
+
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    arguments.asSequence().map { it.toProto(userDataReader) }
+}
+
+internal sealed class GenericArg {
+  companion object {
+    fun from(arg: Any?): GenericArg =
+      when (arg) {
+        is AggregateFunction -> AggregateArg(arg)
+        is Ordering -> OrderingArg(arg)
+        is Map<*, *> ->
+          MapArg(arg.asIterable().associate { (key, value) -> key as String to from(value) })
+        is List<*> -> ListArg(arg.map(::from))
+        else -> ExprArg(Expr.toExprOrConstant(arg))
+      }
+  }
+  abstract fun toProto(userDataReader: UserDataReader): Value
+
+  data class AggregateArg(val aggregate: AggregateFunction) : GenericArg() {
+    override fun toProto(userDataReader: UserDataReader) = aggregate.toProto(userDataReader)
+  }
+
+  data class ExprArg(val expr: Expr) : GenericArg() {
+    override fun toProto(userDataReader: UserDataReader) = expr.toProto(userDataReader)
+  }
+
+  data class OrderingArg(val ordering: Ordering) : GenericArg() {
+    override fun toProto(userDataReader: UserDataReader) = ordering.toProto(userDataReader)
+  }
+
+  data class MapArg(val args: Map<String, GenericArg>) : GenericArg() {
+    override fun toProto(userDataReader: UserDataReader) =
+      encodeValue(args.mapValues { it.value.toProto(userDataReader) })
+  }
+
+  data class ListArg(val args: List<GenericArg>) : GenericArg() {
+    override fun toProto(userDataReader: UserDataReader) =
+      encodeValue(args.map { it.toProto(userDataReader) })
+  }
+}
+
+internal class DatabaseSource
+@JvmOverloads
+internal constructor(options: InternalOptions = InternalOptions.EMPTY) :
+  Stage<DatabaseSource>("database", options) {
+  override fun self(options: InternalOptions) = DatabaseSource(options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> = emptySequence()
+}
+
+class CollectionSource
```

Comment:
in node, i found that i liked calling all of these *StageOptions. It gave consistency and meaning across the Pipeline api. Maybe you would consider calling these *Stage. 

E.g. CollectionStage or CollectionSourceStage, DatabaseStage, LimitStage, WhereStage, ...